src/Common/ThreadPool.h

```diff
@@ -230,15 +232,30 @@ class ThreadFromGlobalPool : boost::noncopyable
         if (!state)
             return false;
         /// Thread cannot join itself.
-        if (*thread_id == std::this_thread::get_id())
+        if (thread_id->id == std::this_thread::get_id())
             return false;
         return true;
     }
 
 private:
     /// The state used in this object and inside the thread job.
     std::shared_ptr<Poco::Event> state;
-    std::shared_ptr<std::thread::id> thread_id;
+
+    struct ThreadIdHolder
+    {
+        /// Should be atomic() because of possible concurrent access between
+        /// assignment and joinable() check.
+        std::atomic<std::thread::id> id;
+    };
+    std::shared_ptr<ThreadIdHolder> thread_id;
```

Comment:
Now I think that maybe it would be better to have `std::atomic<std::thread::id> id` inside state.
Just, why do we need two shared pointers if we can have one? :)

Reply:
Indeed! Fixed and rebased.